### ISSUE TITLE
Support reading files from templates

### ISSUE THREAD

**Issue by omallo** [microconfig/microconfig#44]
It would be very convenient to have a "readFile" custom function which can be used inside templates. The main use case is combining "readFile" with the already existing "base64" custom function in order to load a base64 encoded content from a file into a template.

Some aspects which I think are important for such a "readFile" function:
- It should only accept absolute file paths to avoid additional complexity in figuring out what is the "current" directory.
- It should work with binary files (the "base64" function currently expects a String which might not be ideal).
- It should not fail if the file does not exist but instead allow falling back to a default value.

**Comment by alxmatorin** [comment]
If you use jar version of mc, you can write any Java code using the expression language. 
For readFile you can write:
```
base64Content: #{#base64(T(java.nio.file.Files).readString(T(java.nio.file.Path).of('${this@fileName}')))}
#var fileName:/somePath/someFileName.txt
```
(java11 required cause of Files.readString)

It's a little verbose. If you need to use readFile function regularly, we can add it as a custom function #{readFile('path')}.

**Comment by omallo** [comment]
Thanks for the feedback!

I didn't know that I can use the full expression language, thanks for letting me know. The syntax looks very familiar to me from Spring :)

However, I'm using the native Microconfig binary which is more convenient for us to avoid having any dependency on the Java runtime. So yes, it would be useful to have that functionality also available there if it makes sense to you. Thanks.

**Comment by alxmatorin** [comment]
```
public static byte[] readBytes(String path) //throws exception if the file does not exist

public static byte[] readBytesOrEmpty(String path) //returns empty byte[] if the file does not exist

public static String readString(String path)  //throws exception if the file does not exist

public static String readStringOrEmpty(String path) //returns empty string if the file does not exist
```

are these methods ok for you?

**Comment by omallo** [comment]
@alxmatorin , these look good to me, thanks. I like that you made separate methods for mandatory/optional files to make this more explicit.

**Comment by alxmatorin** [comment]
@omallo released 4.4.1
Check it please

**Comment by omallo** [comment]
@alxmatorin , I tested it with the MacOS binary v4.4.1 which I installed using brew and there it does not work. With the Java version it does work. It's maybe because the new `CustomIoApi` class is not listed in the `reflect-config.json`?

I'm testing with the following simple test configuration:
```
test=#{#base64(#readBytes("/tmp/test.txt"))}
```

I'm getting the following error:
```
Can't evaluate #{#base64(#readBytes("/tmp/test.txt"))}
Cause: EL1006E: Function 'readBytes' could not be found
```

Let me know in case you need more information.

**Comment by omallo** [comment]
I did some further tests with the Java version to test the actual functionality:
```
test1=#{#base64(#readBytes("/tmp/test.txt"))}
test2=#{#base64(#readBytesOrEmpty("/tmp/does-not-exist.txt"))}
test3=#{#readString("/tmp/test.txt")}
test4=#{#readStringOrEmpty("/tmp/does-not-exist.txt")}
test5=#{#base64(#readString("/tmp/test.txt"))}
```

The tests 1-4 with existing and non-existing files work as expected. The test 5 leads to the following error:
```
Can't evaluate #{#base64(#readString("/tmp/test.txt"))}
Cause: EL1001E: Type conversion problem, cannot convert from java.lang.String to byte[]
```
This might be because the `base64` method is overloaded but I'm not totally sure. It's possible to workaround this by invoking `String#getBytes()` inside the expression.

**Comment by alxmatorin** [comment]
@omallo 
spring EL doesn't support method overloading, I'll rename base64(byte[]) to bytesBase64(bytes[]).
Thank you for for catching that :)

**Comment by alxmatorin** [comment]
@omallo released 4.4.2

**Comment by alxmatorin** [comment]
@omallo released 4.4.3

**Comment by omallo** [comment]
@alxmatorin , I tested again with the version 4.4.3 and now everything seems to work fine with the binary distribution, thanks!
